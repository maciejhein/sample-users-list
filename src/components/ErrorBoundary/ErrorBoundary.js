import React, { Component } from "react";
import PropTypes from "prop-types";

class ErrorBoundary extends Component {
  state = {
    hasError: false,
    error: null,
    errorInfo: null,
  };

  componentDidCatch(error, errorInfo) {
    this.setState({ hasError: true, error, errorInfo });
  }
  // TODO we need to create custom component Error
  render() {
    const { error, errorInfo, hasError } = this.state;

    if (hasError) {
      return (
        <div>
          <span>{error.toString()}</span>
          <span>{errorInfo.componentStack}</span>
        </div>
      );
    }

    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ErrorBoundary;
