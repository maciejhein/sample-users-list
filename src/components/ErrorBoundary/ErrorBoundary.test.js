import React from "react";
import { render } from "@testing-library/react";

import ErrorBoundary from "./ErrorBoundary";

describe("COMPONENT - ErrorBoundary", () => {
  let consoleErrorSpy;

  // Dummy components, for testing purposes
  function ChildComponent() {
    return <p>I'm child component!</p>;
  }

  function ComponentWithError() {
    throw new Error("Something went wrong...");
  }

  beforeAll(() => {
    consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {
      // Allows suppressing error bounadry logs
    });
  });

  afterAll(() => {
    consoleErrorSpy.mockRestore();
  });

  it("renders the child component if there is no error", () => {
    const { asFragment } = render(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders fallback Error component if there is error", () => {
    const { getByText } = render(
      <ErrorBoundary>
        <ComponentWithError />
      </ErrorBoundary>
    );

    expect(getByText("Error: Something went wrong...")).toBeTruthy();
  });
});
