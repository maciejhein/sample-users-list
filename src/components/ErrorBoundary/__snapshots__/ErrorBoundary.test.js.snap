// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - ErrorBoundary renders the child component if there is no error 1`] = `
<DocumentFragment>
  <p>
    I'm child component!
  </p>
</DocumentFragment>
`;
